### PR TITLE
Pull the shared python modules out of exchange and put them in their own contract family

### PR DESCRIPTION
**Pull request by cmickeyb**
Split out the "contracts" module from the exchange contract and put it into a new "common" contract family. While strictly speaking, this is not a contract family, the build processes incorporate module creation very easily. In addition, this provides a framework for adding common functionality (like wrappers for tests, shared methods, etc).

### CODE REVIEW

**Review by g2flyer — COMMENTED** [review]
It's a bit hard to see what exactly is outsourced as the reusable abstraction, i guess partially as there is only a single family here. I noticed there is a separate PR #63 with a new family but as there is no before/after that doesn't really help.  All that said, refactoring out common functionality is always good,  i certainly have no objection to the changes.

The (docker) tests still build and run mostly successfully but fail eventually in below: is that expected? 

```
.... 
contracts_container  | 5:  gs_start.sh: start guardian_service
contracts_container  | 5: /project/pdo/run/opt/pdo/contracts/inference/scripts/gs_start.sh: line 90:  2015 Aborted                 (core dumped) ${F_SERVICE_CMD} $@ 2> $EFILE > $OFILE
contracts_container  | 5:  gs_start.sh: ERROR: guardian service failed to start
contracts_container  | 5:  script_test.sh: ERROR: operation failed: /project/pdo/run/opt/pdo/contracts/inference/scripts/gs_start.sh -c -o /project/pdo/run/opt/pdo/logs -- --loglevel debug --config guardian_service.toml --config-dir /project/pdo/run/opt/pdo/etc/contracts --identity guardian_service --bind host localhost --bind service_host localhost
contracts_container  | 5:  script_test.sh: shutdown guardian and storage service
contracts_container  | 5: /project/pdo/run/opt/pdo/contracts/inference/scripts/gs_stop.sh: line 37: kill: (2015) - No such process
contracts_container  | Errors while running CTest
contracts_container  | Output from these tests are in: /project/pdo/contracts/build/Testing/Temporary/LastTest.log
contracts_container  | Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
contracts_container  | 5/5 Test #5: system-inference-script ...........***Failed   35.70 sec
contracts_container  |
contracts_container  | 60% tests passed, 2 tests failed out of 5
contracts_container  |
contracts_container  | Total Test time (real) = 812.19 sec
contracts_container  |
contracts_container  | The following tests FAILED:
contracts_container  |    2 - system-digital-asset-functional (Failed)
contracts_container  |    5 - system-inference-script (Failed)
contracts_container  | make[1]: Leaving directory '/project/pdo/contracts/build'
contracts_container  | make[1]: *** [Makefile:71: test] Error 8
contracts_container  | make: *** [Makefile:49: test] Error 2
contracts_container  | make: Leaving directory '/project/pdo/contracts'
contracts_container  |  run_contracts_tests.sh: ERROR: operation failed: make -C /project/pdo/contracts TEST_LOG_LEVEL=warn TEST_SERVICE_HOST=localhost TEST_LEDGER=http://127.0.0.1:6600 test
contracts_container exited with code 111
Aborting on container exit...
```

**Comment by cmickeyb** [comment]
> It's a bit hard to see what exactly is outsourced as the reusable abstraction, i guess partially as there is only a single family here. I noticed there is a separate PR #63 with a new family but as there is no before/after that doesn't really help. All that said, refactoring out common functionality is always good, i certainly have no objection to the changes.

This is mostly about moving the pdo.contracts.* python modules out into their own directory. There are two modules that matter, one for common routines for context and the other a suite of jupyter widgets (with the corresponding documents that expose those widgets). The README has a little information but definitely needs to be improved. 

> The (docker) tests still build and run mostly successfully but fail eventually in below: is that expected?

These should not fail (and don't seem to be failing for me). Can you post the logs from the full test run?

**Comment by g2flyer** [comment]
> > The (docker) tests still build and run mostly successfully but fail eventually in below: is that expected?
> 
> These should not fail (and don't seem to be failing for me). Can you post the logs from the full test run?

Here's the output from `make -C docker test` [make.test.pr-64-fails.log](https://github.com/user-attachments/files/17564621/make.test.pr-64-fails.log).   Issue could be maybe the low-powered NUC?  I'm currently re-running the tests from `main` branch to see whether i get the same problem (and so might be issue with my machine?) ...

**Comment by g2flyer** [comment]
> I'm currently re-running the tests from `main` branch to see whether i get the same problem (and so might be issue with my machine?) ...

Seems i get problems also when running from main: first it actually didn't even get really going as CCF couldn't be properly started (endless "waiting" ...) and then i got again a gs_start issue as before.   I also thought first it might be related to configs in `docker/make.loc` but i got these problem now even after commenting them out.   So if you don't get any problems on your system, then maybe it is an issue with under-powered gemini lake NUCs?

**Comment by cmickeyb** [comment]
> Seems i get problems also when running from main: first it actually didn't even get really going as CCF couldn't be properly started (endless "waiting" ...) and then i got again a gs_start issue as before. I also thought first it might be related to configs in `docker/make.loc` but i got these problem now even after commenting them out. So if you don't get any problems on your system, then maybe it is an issue with under-powered gemini lake NUCs?

I've run this several times without issues. Maybe we should first push the test workflow that i have ready to go. its really slow, but does a completely clean build and run. Are you running PDO in HW mode? Although the errors in the earlier run appeared to be outside the enclave (gs-start.sh). also... if you can capture the logs that would be very helpful.

File: common-contract/MANIFEST [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+setup.py
+pdo/__init__.py
+pdo/contracts/common.py
+pdo/contracts/__init__.py
+pdo/contracts/jupyter/services.py
+pdo/contracts/jupyter/groups.py
+pdo/contracts/jupyter/common_widgets.py
+pdo/contracts/jupyter/utility.py
+pdo/contracts/jupyter/keys.py
+pdo/contracts/jupyter/__init__.py
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply common-contract/MANIFEST]: Fixed.

File: common-contract/MANIFEST.in [hunk 2]
```diff
@@ -0,0 +1,4 @@
+recursive-include ../build/example-contract *.b64
+recursive-include etc *.toml
+recursive-include context *.toml
+recursive-include scripts *.psh
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply common-contract/MANIFEST.in]: Fixed

**Comment by g2flyer** [comment]
> > Here's the output from `make -C docker test` [make.test.pr-64-fails.log](https://github.com/user-attachments/files/17564621/make.test.pr-64-fails.log). Issue could be maybe the low-powered NUC? I'm currently re-running the tests from `main` branch to see whether i get the same problem (and so might be issue with my machine?) ...
> 
> I would guess that both tests are failing because of LMDB (one for sure). That may very well be a resource problem on the NUC. Both of the tests that fail use the blockstore to send large objects to the contract enclave. The digital asset test is definitely failing during that transfer. And, given that the guardian service uses the same method for moving big images, I would guess it is also failing for that reason.

I have 264G available free space but i remember lmdb giving problems in docker due to sparse-files being fully allocated.  I'm currently running now in wsl on my laptop (which has 138GB free space) to see whether i run in same issue.  If it is really disk space and 264GB is not enough, we might want to reconsider lmdb sparse file config?

> I'm thinking that the right solution is to enable workflow_dispatch version of the tests on github. thoughts?

Enabling that certainly seems a good idea. If that would work that of course would be a work-around for the "low-disk" folks, although debuging anything there would be very painful (e.g., no way to inspect container state via a shell?)

**Comment by cmickeyb** [comment]
Just to close on the errors being seen. 

1) Adding locks to the keyvalue and blockmanager classes in the PDO code addressed lmdb failures. Will pull the new PDO version into contracts.
2) The issue with the guardian service appears to be that the NUC used for testing does not support AVX operations required by tensorflow. This error is "to be expected".

**Review by g2flyer — APPROVED** [review]
With the new locking, test do now work successfully on WSL on my laptop where it previously failed.  Didn't retest on FLC NUCs as they will fail due to HW-limitations (AVX missing).

**Pull request merged**